crates/module-system/module-implementations/sov-evm/src/rpc/mod.rs

```diff
@@ -357,19 +340,44 @@ where
     #[rpc_method(name = "eth_estimateGas")]
     pub fn eth_estimate_gas(
         &self,
-        _request: TransactionRequest,
-        _block_number: Option<String>,
-        _state: &mut ApiStateAccessor<S>,
+        request: TransactionRequest,
+        block_number: Option<String>,
+        state: &mut ApiStateAccessor<S>,
     ) -> RpcResult<U64> {
-        // TODO EVM: #1510
-        Ok(U64::from(100_000))
+        debug!("EVM module JSON-RPC request to `eth_estimateGas`");
+        let result = self.call(request, block_number, state)?;
+        let gas_used = result.gas_used();
+        const ABSOLUTE_MARGIN: f64 = 1.5;
+        const RELATIVE_MARGIN: u64 = 100_000;
+        #[allow(clippy::float_arithmetic)]
```

Comment:
We can't use float_arithmetic because it's not fully deterministic. Instead, let's use checked u64 division and multiplication.